### PR TITLE
T2 h 119 migrate updates to delta model

### PR DESCRIPTION
**Pull request by emargetis**
### Delta Updates
The current system for sending state updates from the server to the client sends the entire state of the server to every client each time a change (mutation) is made on the server state. This can be an expensive operation, especially if the state object is large since it will need to be sent over the websocket connection and then parsed / applied by each client each time.

As an alternative, this PR implements a Delta Update system. The delta update system consists of three operations that the server can send to the client after each mutation to the server's state:
- `put`: tells the client to replace their existing value of a given key (or create a new one) with the new value in its local storage
- `del`: tells the client to remove the key and its value from its local storage
- `clear`: tells the client to clear its entire local storage

These operations are grouped into a patch array by the server after each mutation occurs and it is then broadcasted along with some other data to each of the clients. Once each client receives the object, it applies the patch array locally to match the server state.

### ClientID and ULID
In addition to implementing delta updates as outlined above, the server now keeps track of the latest mutation received by each client, updates that information as new mutations come in, and then sends that information back along with each patch message that is broadcasted. This information allows each client to remove any pending transactions from its queue so that it does replay them on top of their local state or attempt to resend them to the server after they have been acknowledge and incorporated by the server. In order to do this, we implemented a `clientID` as a property on the saurus object that is sent along with every mutation from the client to the server. Additionally, we changed the transaction ID that is sent with every mutation from the client to the server to be a ULID (as opposed to a timestamp) so that if multiple mutations occur within the same millisecond, they can be ordered for quick removal from the pending queue after a broadcast message is received from the server.

### Further Development
As of now, every single mutation across every client results in a broadcasted message with a patch object. However, the broadcast messages could batch multiple patch messages to reduce the total number of messages being sent over the websocket connection and reduce the fixed overhead associated with processing messages on both the server and client side. This batching approach could also be applied to mutations when they are sent from the client to the server, if sending every mutation separately becomes a bottleneck.

### CODE REVIEW

File: do-server/src/index.mjs [hunk 1]
```diff
@@ -1,19 +1,36 @@
...
   }
 
   set(key, value) {
+    this.patch.push({
+      op: 'put',
+      key,
+      value,
+    });
+
     this.canon[key] = value;
```
Comment (FrancoWaisfeld) [inline do-server/src/index.mjs:23]: Could this potentially go above pushing the object into the patch array?

Comment (FrancoWaisfeld) [inline do-server/src/index.mjs]: Could this potentially go above pushing the object into the patch array? Same goes for the delete method below.

**Review by FrancoWaisfeld — CHANGES_REQUESTED** [review]

File: do-server/src/index.mjs (hunk 1 above)
Comment (emargetis) [reply do-server/src/index.mjs]: Good call! We probably want to make sure the values get changed in the sever state before putting them in the patch array just to ensure the client never gets ahead of the server. Will fix now.

Comment (emargetis) [reply do-server/src/index.mjs]: new commit pushed

**Comment by sklarfox** [comment]
This is a great update! It's awesome we don't have to maintain duplicate `canonState` and `localState` on the client side, that seems like a big optimization as the storage files get larger. 🚢

**Review by FrancoWaisfeld — APPROVED** [review]
Everything looks good to me now!

**Pull request merged**